libs/execution_context/ebpf_native.c

```diff
@@ -0,0 +1,1219 @@
+    uint16_t service_name_length,
+    _In_ const GUID* module_id,
+    _Out_ size_t* count_of_maps,
+    _Out_ size_t* count_of_programs)
+{
+    ebpf_result_t result;
+    ebpf_lock_state_t hash_table_state = 0;
+    ebpf_lock_state_t state = 0;
+    bool lock_acquired = false;
+    ebpf_native_t* native_module = NULL;
+    ebpf_native_t** existing_native_module = NULL;
+    wchar_t* local_service_name = NULL;
+
+    local_service_name = ebpf_allocate((size_t)service_name_length + 2);
+    if (local_service_name == NULL) {
+        result = EBPF_NO_MEMORY;
+        goto Done;
+    }
+    memcpy(local_service_name, (uint8_t*)service_name, service_name_length);
+
+    ebpf_native_load_driver(local_service_name);
+
+    // Find the native entry in hash table.
+    hash_table_state = ebpf_lock_lock(&_ebpf_native_client_table_lock);
+    lock_acquired = true;
+    result =
+        ebpf_hash_table_find(_ebpf_native_client_table, (const uint8_t*)module_id, (uint8_t**)&existing_native_module);
+    if (result != EBPF_SUCCESS) {
+        result = EBPF_OBJECT_NOT_FOUND;
+        goto Done;
+    }
+    native_module = *existing_native_module;
+    state = ebpf_lock_lock(&native_module->lock);
+    if (native_module->initialized) {
+        // This client has already been initialized.
+        result = EBPF_OBJECT_ALREADY_EXISTS;
+        ebpf_lock_unlock(&native_module->lock, state);
+        goto Done;
+    }
+    if (native_module->detaching || native_module->unloading) {
+        // This client is already detaching / unloading.
+        result = EBPF_EXTENSION_FAILED_TO_LOAD;
+        ebpf_lock_unlock(&native_module->lock, state);
+        goto Done;
+    }
+    native_module->initialized = true;
+    native_module->service_name = local_service_name;
+    ebpf_lock_unlock(&native_module->lock, state);
+
+    // Get map and program count;
+    *count_of_maps = _ebpf_native_get_count_of_maps(native_module);
+    *count_of_programs = _ebpf_native_get_count_of_programs(native_module);
+
+Done:
+    if (lock_acquired) {
+        ebpf_lock_unlock(&_ebpf_native_client_table_lock, hash_table_state);
+        lock_acquired = false;
+    }
+    if (result != EBPF_SUCCESS) {
+        ebpf_free(local_service_name);
+    }
+
+    return result;
+}
+
+ebpf_result_t
+ebpf_native_load_programs(
+    _In_ const GUID* module_id,
+    size_t count_of_map_handles,
+    _Out_writes_(count_of_map_handles) ebpf_handle_t* map_handles,
+    size_t count_of_program_handles,
+    _Out_writes_(count_of_program_handles) ebpf_handle_t* program_handles)
+{
+    ebpf_result_t result;
+    ebpf_lock_state_t state = 0;
+    ebpf_lock_state_t native_state = 0;
+    bool lock_acquired = false;
+    bool native_lock_acquired = false;
+    ebpf_native_t** existing_native_module = NULL;
+    ebpf_native_t* native_module = NULL;
+    wchar_t* local_service_name = NULL;
+
+    // Find the native entry in hash table.
+    state = ebpf_lock_lock(&_ebpf_native_client_table_lock);
```

Comment:
What is this lock protecting? ebpf_hash_table is lock-free and shouldn't need this.

Reply:
Discussed offline. The table lock is needed to ensure one thread does not cleanup the native module object while the other thread has queried the object from hash table and trying to access it.